### PR TITLE
NHWC batch-norm forward complete

### PR DESCRIPTION
**Pull request by qianfengz**
Following P.R. 182.  This P.R includes
1) Some necessary change to the host common utility
2) Implementation of `NHWC  bnorm-fwd` using `reduce_multiblock` method (kernels, device and test example)
3) Instance library for `bnorm-fwd` (support both bnorm-fwd using `blockwise` and `multiblock` reduction)
4) Profiler module for `bnorm-fwd` (support both `bnorm-fwd` using `blockwise` and `multiblock` reduction)
5) CTest for bnorm-fwd (support both `bnorm-fwd` using `blockwise` and `multi-block` reduction)

### CODE REVIEW

File: example/19_bnorm_fwd/bnorm_fwd_nhwc_blockwise.cpp [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+    1,   // KThreadSliceSize,
+    2,   // InOutVectorSize,
+    2>;  // ScaleBiasMeanVarVectorSize
+
+static struct option long_options[] = {{"inOutLengths", required_argument, nullptr, 'D'},
+                                       {"verify", required_argument, nullptr, 'v'},
+                                       {"help", no_argument, nullptr, '?'},
+                                       {nullptr, 0, nullptr, 0}};
+
+class SimpleAppArg
```
Comment (zjing14) [inline example/19_bnorm_fwd/bnorm_fwd_nhwc_blockwise.cpp:47]: example of bnorm_fwd_nhwc_blockwise/multiblock share a lot of code, could we extract common codes into a head file?

**Comment by zjing14** [comment]
Could we treat blockwise as a special case of multiblock, and unify the two implementations?

File: example/19_bnorm_fwd/bnorm_fwd_nhwc_blockwise.cpp (hunk 1 above)
Comment (qianfengz) [reply example/19_bnorm_fwd/bnorm_fwd_nhwc_blockwise.cpp:47]: The same/similar names,  but different codes in different cpp files

**Comment by qianfengz** [comment]
> Could we treat blockwise as a special case of multiblock, and unify the two implementations?

This could not be done.   `multi-block` is needed when the size of reduced dimensions is much bigger than the size of invariant dimensions, for example when `N*H*W` is much bigger than `C`,   we need to use `multi-block` to completely utilize all the compute-units.  And `multi-block` either use atomic-add or second-call to finish the final reduction. In other hand, `blockwise` can finish the reduction in single call

**Comment by asroy** [comment]
included by #211

**Pull request closed without merging**